src/workerd/api/eventsource.c++

```diff
@@ -0,0 +1,520 @@
+#include <workerd/io/features.h>
+#include <workerd/jsg/exception.h>
+#include <workerd/util/mimetype.h>
+
+namespace workerd::api {
+
+namespace {
+class EventSourceSink final: public WritableStreamSink {
+public:
+  EventSourceSink(EventSource& eventSource) : eventSource(eventSource) {}
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    // The event stream is a new-line delimited format where each line represents an event.
+    // We need to scan the buffer for end-of-line characters. When we find one, everything before
+    // it is pushed into the event queue and we keep scanning. If we do not find an end-of-line
+    // sequence in the remaining input, we buffer it and wait for the next write to continue
+    // scanning, or until the stream is ended or aborted.
+
+    if (eventSource == kj::none) {
+      // Write was received after end() or abort() was called.
+      // We'll just ignore the write.
+      return kj::READY_NOW;
+    }
+
+    kj::ArrayPtr<const char> input(static_cast<const char*>(buffer), size);
+
+    // The stream may or may not begin with the UTF-8 BOM (%xFEFF). If this is the
+    // first write, we need to check for it and skip it if it is present.
+    // The BOM is a 3-byte sequence (0xEF, 0xBB, 0xBF) that encodes the Unicode
+    // codepoint U+FEFF. We only want to check for this once.
+    if (!bomChecked) {
+      bomChecked = true;
+      if (size >= 3 && input[0] == '\xEF' && input[1] == '\xBB' && input[2] == '\xBF') {
+        input = input.slice(3);
+      }
+    }
+
+    while (input != nullptr) {
+      KJ_IF_SOME(found, findEndOfLine(input)) {
+          auto prefix = kept.releaseAsArray();
+          // Feed the line into the processor.
+          feed(kj::str(prefix, input.slice(0, found.pos)));
+          input = found.remaining;
+          // If we've reached the end of the input, input will == nullptr here.
+      } else {
+        // No end-of-line found, buffer the input.
+        kept.addAll(input.begin(), input.end());
+        input = nullptr;
+      }
+    }
+
+    // Release any buffered events to the EventSource
+    release();
+
+    return kj::READY_NOW;
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const kj::byte>> pieces) override {
+    for (auto& piece : pieces) {
+      co_await write(piece.begin(), piece.size());
+    }
+    co_return;
+  }
+
+  kj::Promise<void> end() override {
+    // The stream has finished. There's really nothing left to do here. Any partially
+    // filled data will be dropped on the floor.
+    clear();
+    return kj::READY_NOW;
+  }
+
+  void abort(kj::Exception reason) override {
+    // There's really nothing to do here.
+    clear();
+  }
+
+private:
+  kj::Maybe<EventSource&> eventSource;
+  // Retained bytes to be processed in the next write.
+  kj::Vector<char> kept;
+  // The collected messages that are pending to be dispatched as events
+  kj::Vector<EventSource::PendingMessage> pendingMessages;
+
```

Comment:
Super nit; New lines are separating some of the fields but not all. Maybe we don't need them?